lib/Mojolicious/Plugin/OpenAPI/Cors.pm

```diff
@@ -48,63 +169,187 @@ Mojolicious::Plugin::OpenAPI::Cors - OpenAPI plugin for Cross-Origin Resource Sh
+L</openapi.cors_exchange>:
+
+  package MyApp::Controller::User;
+
+  sub get_user {
     # Validate incoming CORS request with _validate_cors()
-    my $c = shift->openapi->cors_simple("_validate_cors")->openapi->valid_input or return;
+    my $c = shift->openapi->cors_exchange("_validate_cors")->openapi->valid_input or return;
 
+    # Will only run this part if both the cors_exchange and valid_input was
+    # successful.
     $c->render(openapi => {user => {}});
   }
 
+  # This method must return undef on success. Any true value will be used as
+  # an error.
   sub _validate_cors {
-    my ($c, $args) = @_;
+    my ($c, $params) = @_;
+
+    # $params->{type} is set to "preflighted", "real" or "simple"
+    $c->app->log->debug("Got CORS $params->{type} request");
+
+    # The following "Origin" header check is the same for both simple and
+    # preflighted.
+    return "/Origin" unless $params->{origin} =~ m!^https?://whatever.example.com!;
 
-    # Check the origin of the request
-    if ($args->{origin} =~ m!^https?://whatever.example.com!) {
+    # The following checks are only valid if preflighted...
 
-      # Setting the "Access-Control-Allow-Origin" will mark this request as valid
-      $c->res->headers->header("Access-Control-Allow-Origin" => $args->{origin});
+    # Check the Access-Control-Request-Headers header
+    return "Bad stuff." if $params->{headers} and $params->{headers} =~ /X-No-Can-Do/;
+
+    # Check the Access-Control-Request-Method header
+    return "Not cool." if $params->{method} and $params->{method} eq "delete";
+
+    # Set the following header for both simple and preflighted on success
+    # or just let the auto-renderer handle it.
+    $c->res->headers->header("Access-Control-Allow-Origin" => $params->{origin});
+
+    # Set Preflighted response headers, instead of using the default
+    if ($params->{type} eq "preflighted") {
+      $c->res->headers->header("Access-Control-Allow-Headers" => "X-Whatever, X-Something");
+      $c->res->headers->header("Access-Control-Allow-Methods" => "POST, GET, OPTIONS");
+      $c->res->headers->header("Access-Control-Max-Age" => 86400);
     }
+
+    # Return undef on success.
+    return undef;
   }
 
 =head1 DESCRIPTION
 
-L<Mojolicious::Plugin::OpenAPI::Cors> is a plugin for accepting Simple
-Cross-Origin Resource Sharing requests, by looking at the "Origin" header. See
+L<Mojolicious::Plugin::OpenAPI::Cors> is a plugin for accepting Preflighted or
+Simple Cross-Origin Resource Sharing requests. See
 L<https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS> for more details.
 
 This plugin is loaded by default by L<Mojolicious::Plugin::OpenAPI>.
 
-Note that this plugin currently EXPERIMENTAL! Please let me know if you have
-any feedback.
+Note that this plugin currently EXPERIMENTAL! Please comment on
+L<https://github.com/jhthorsen/mojolicious-plugin-openapi/pull/102> if
+you have any feedback or create a new issue.
+
+=head1 STASH VARIABLES
```

Comment:
I especially like the stash var interface here. I wonder if you could go further with it and allow the exchange callback to be defined here too? Then it just gets run with the regular validators?

Reply:
I don't understand. Isn't that what "openapi_cors_default_exchange_callback" is?